andela/ah-robotics
# #162949075 Users should be able to create their profiles

### Getting Started

1. Open terminal and clone the repo:

    `git clone https://github.com/andela/ah-robotics.git`

2. Change directory into the cloned repository.

    `cd ah-robotics`

3. Checkout to the Feature branch

    `git checkout ft-user-profile-162949075`

4. Create a virtual environment by running the following commands:

    `virtualenv -p python3 env`

    or

    `python3 -m virtualenv env`

5. Once the virtual environment has been setup, activate it and install the packages required by the project,
    these are contained in the file named `requirements.txt`

      `source env/bin/activate`

      `pip install -r requirements.txt`

### setting up the database

> Ensure PostgreSQL is installed and is running psql, follow these steps to create a new user and a database which will be used in the application.

  1. Start the postgres database server using the command:
  
      `pg_ctl -D /usr/local/var/postgres start`

  2. Run psql interactive terminal

      `psql postgres`

  3. Create a new user:

      `CREATE USER sample_username WITH PASSWORD 'sample_password';`

  4. Grant privileges to the user:

      `ALTER USER sample_username CREATEDB;`

  5. Create a database:

      `CREATE DATABASE sample_database_name WITH OWNER sample_username;`
       
  6.   Exit the psql shell using `\q`


  7. Create a .env as guided by the .env_example and add the following configuration variables

      ```source-json
      source env/bin/activate
      export DB_NAME="database_name"
      export DB_USER="sample_username"
      export DB_PASS="sample_password"
      export DB_HOST="host_name_eg_localhost"
      export DB_PORT="port_number_eg_5432"
      export TIME_DELTA="30"
      export EMAIL_HOST="smtp.gmail.com"
      export EMAIL_HOST_USER="user@example.com"
      export EMAIL_HOST_PASSWORD="email_password"
      export EMAIL_PORT=587
      export PASSWORD_RESET_URL="api/v1/reset_password"
      export CLIENT_DOMAIN="127.0.0.1:8000"
      export CLIENT_DOMAIN="127.0.0.1:8000"
      export CLOUDINARY_NAME="your_cloudinary_name"
      export CLOUDINARY_KEY="your_cloudinary_key"
      export CLOUDINARY_SECRET="your_secret_key"
      ```
  8. Source .env to set the configuration variables:

      `source .env`
  9. Run migrations to create the tables in the database:

      ```source-json
      python manage.py makemigrations
      python manage.py migrate
      ```

---
### Collect static files

Run the following command to collect static files.

`python manage.py collectstatic`

### Run The Server

Once everything has been setup, to run the application run the command:

   `python manage.py runserver`

The application can be accessed in the url <http://localhost:8000>

---

### Testing the feature

1. Run the server from the terminal using python manage.py runserver.

2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL <http://127.0.0.1:8000/api/v1/users/> to register a user.
3. On the body section on postman select `raw` and type as `JSON/application`.
Add your payload to the body section using the following format `but NOT with the exact credentials`:

> Kindly use a genuine email you have access to because an email will be sent, do NOT use a real password.

   ```
    http://127.0.0.1:8000/api/v1/users/
 ```
        {
        "user": {
            "username": "dojo1234",
            "email": "user@example.com",
            "password": "test1234#"
            }
          }
     
4. Once all the fields are inserted, click POST to register the user successfully.
5. Repeat this step again to register another user.
6. Change the endpoint to <http://127.0.0.1:8000/api/v1/users/login/>
7. On the body section on postman select `raw` and type as `JSON/application`. Add the following details with the credentials you registered a user with:

    ```
     http://127.0.0.1:8000/api/v1/users/login/
   ```

    ```
        {
        "user": {
            "email": "user@example.com",
            "password": "test1234#"
            }
          }
     ```
8. Send the Request and copy the token section of the response gotten.
9. Change to a `GET` request and add the endpoint: <http://127.0.0.1:8000/api/v1/profiles/> to view search bar.
10. Click the tab labeled `Headers` and add information as in the screenshot below. Ensure you type the word token before pasting you token in the authorization row.
> Click the first row and add `content-type`. Match it in adjacent column same row with `application/json`.
> in the second row, type `Authorizations` and in the corresponding field, type `token` then paste your token.
<img width="1039" alt="screenshot 2019-01-24 at 22 18 08" src="https://user-images.githubusercontent.com/14871749/51702885-05395400-2026-11e9-9065-dc3d1489938c.png">
11. To view a specific user use this url format below: 

   
   ``` 
http://127.0.0.1:8000/api/v1/profiles/<sample_user_name/>
 ```
    
12. To update the user profile, change the request to `PATCH`  
   ```
<http://127.0.0.1:8000/api/v1/profiles/sample_user_name/>
```
In the body section use the form data to update fields(bio and image)

<img width="692" alt="screenshot 2019-01-25 at 12 56 40" src="https://user-images.githubusercontent.com/9401930/51740364-db734200-20a4-11e9-988c-3abe27c147d1.png">




### Running tests
1. Go to the terminal and ensure that the server is not running.
2. Run the tests 
 `python manage.py test`

### What are the relevant pivotal tracker stories?
[#162949075](https://www.pivotaltracker.com/n/projects/2235171/stories/162949075)

Screenshots:
1. User Registration triggers profile creation.

<img width="1042" alt="screen shot 2019-01-24 at 19 22 49" src="https://user-images.githubusercontent.com/14871749/51701229-0e282680-2022-11e9-906f-bee3b8c489b6.png">

2. Login the user
<img width="1036" alt="screen shot 2019-01-24 at 19 24 05" src="https://user-images.githubusercontent.com/14871749/51701304-3f085b80-2022-11e9-98b1-4b7f880b97de.png">

3. View all profiles
<img width="1037" alt="screen shot 2019-01-24 at 19 25 49" src="https://user-images.githubusercontent.com/14871749/51701318-4596d300-2022-11e9-9edf-f8d989ac7b01.png">

4. View one profile
<img width="1039" alt="screen shot 2019-01-24 at 19 26 38" src="https://user-images.githubusercontent.com/14871749/51701329-4a5b8700-2022-11e9-9efc-0fa0e349f532.png">

5. Update my profile
<img width="692" alt="screenshot 2019-01-25 at 12 56 40" src="https://user-images.githubusercontent.com/9401930/51740465-2b520900-20a5-11e9-89eb-f96753230a31.png">
